src/libexpr/eval.cc

```diff
@@ -1994,42 +1952,58 @@ void ExprConcatStrings::eval(EvalState & state, Env & env, Value & v)
 
```

Comment:
This PR currently breaks path appending in a significant way, where a `/` isn't necessary for appending anymore for the lazy path root, but it is for subsequent appends:

```
Welcome to Nix 2.12.0pre20221205_fcdca3d. Type :? for help.

nix-repl> lazyPath = (fetchTree { type = "git"; url = ./.; }).outPath 

nix-repl> lazyPath 
/home/tweagysil/src/nix/

nix-repl> lazyPath + "foo"
/home/tweagysil/src/nix/foo

nix-repl> (lazyPath + "foo") + "bar"
/home/tweagysil/src/nix/foobar

nix-repl> (lazyPath + "foo") + "/bar"
/home/tweagysil/src/nix/foo/bar
```

Yes this does make sense to a degree, but it can easily lead to mistakes. My recommendation is to throw an error if the string that is being appended doesn't start with a `/`, but only for lazy paths (because we can't break backwards compatibility for the traditional paths).

Reply:
> My recommendation is to throw an error if the string that is being appended doesn't start with a `/`, but only for lazy paths (because we can't break backwards compatibility for the traditional paths).

Only for lazy path _roots_, because existing "legacy" expressions may be loaded through lazy trees, and should continue to work as usual.

In other words, keep this behavior:

```
nix-repl> (lazyPath + "foo") + "bar"
/home/tweagysil/src/nix/foobar
```

Reply:
It's probably fine to always throw an error, since `lazyPath + "foo"` was never a path that can actually exist. E.g. pre lazy trees we had:
```
nix-repl> lazyPath = (fetchTree { type = "git"; url = /home/eelco/Dev/patchelf; }).outPath

nix-repl> lazyPath + "foo"
"/nix/store/k5gs2alrzw2zz08bp92zd93dsz1n97kl-sourcefoo"
```